### ISSUE TITLE
SerializationException on kernel in distributed mode

### ISSUE THREAD

**Issue by tanoshko** [griddynamics/jagger#842]
# Issue

Kernel fails to start in distributed mode (2 kernels) with error "java.lang.RuntimeException: org.jboss.serial.exception.SerializationException: Could not create instance of com.griddynamics.jagger.user.test.configurations.JTestDefinition$Builder". Please see full stack trace below.

Most probably it happens because JTestDefinition.Builder doesn't have public constructor.

Code sample which fails:

`Configuration
public class ProductsResourceScenario extends JaggerPropertiesProvider {
    private static final Id testGroupId = Id.of("Products Resource");
    private static final int DURATION_IN_MINUTES = 1;
    private static final int PRODUCTS_PER_REQUEST = 1;
    private static final int REQUESTS_PER_SECOND = 1;

    Value("${fcc.endpoints.products.resource}")
    private String[] fccEndpoints;

    private JLoadScenario loadScenario(String testName, Map<String, String> queryParams) {
        JTestDefinition testDef = JTestDefinition
                .builder(Id.of(testName), new FccEndpointsProvider(fccEndpoints))
                .withComment("Get Products test")
                .withInvoker(new GetProductsInvokerProvider())
                .withQueryProvider(new ProductIdsFromFccQueryProvider(ACTIVE, PRODUCTS_PER_REQUEST, fccEndpoints[0], queryParams))
                .addValidator(JHttpResponseStatusValidatorProvider.of(200))
                .build();

        JLoadProfile loadProfile = JLoadProfileRps.builder(RequestsPerSecond.of(REQUESTS_PER_SECOND)).build();
        JTerminationCriteria terminationCriteria = JTerminationCriteriaDuration.of(DurationInSeconds.of(DURATION_IN_MINUTES * 60));

        JLoadTest test1 = JLoadTest.builder(Id.of(testName), testDef, loadProfile, terminationCriteria).build();

        JParallelTestsGroup testGroup = JParallelTestsGroup.builder(testGroupId, test1).build();

        return JLoadScenario.builder(Id.of(testName), testGroup).build();
    }

    Bean
    public JLoadScenario expandReviews() {
        Map<String, String> queryParams = new HashMap<>();
        queryParams.put("_expand", "reviews");

        return loadScenario("productsResource_ExpandReviews", queryParams);
    }
}`

# Steps to reproduce

1. Run Jagger in distributed mode 

> ERROR 11:20:30,957 [Thread-0-EventThread] ClientCnxn | Error while calling watcher 
> java.lang.RuntimeException: org.jboss.serial.exception.SerializationException: Could not create instance of com.griddynamics.jagger.user.test.configurations.JTestDefinition$Builder - com.griddynamics.jagger.user.test.configurations.JTestDefinition$Builder
>         at com.griddynamics.jagger.util.SerializationUtils.deserialize(SerializationUtils.java:157)
>         at com.griddynamics.jagger.coordinator.zookeeper.ZNodeImpl.getObject(ZNodeImpl.java:73)
>         at com.griddynamics.jagger.coordinator.zookeeper.ZookeeperCoordinator.getEntries(ZookeeperCoordinator.java:268)
>         at com.griddynamics.jagger.coordinator.zookeeper.ZookeeperCoordinator.access$300(ZookeeperCoordinator.java:57)
>         at com.griddynamics.jagger.coordinator.zookeeper.ZookeeperCoordinator$2.process(ZookeeperCoordinator.java:214)
>         at org.apache.zookeeper.ClientCnxn$EventThread.run(ClientCnxn.java:488)
> Caused by: org.jboss.serial.exception.SerializationException: Could not create instance of com.griddynamics.jagger.user.test.configurations.JTestDefinition$Builder - com.griddynamics.jagger.user.test.configurations.JTestDefinition$Builder
>         at org.jboss.serial.classmetamodel.ClassMetaData.newInstance(ClassMetaData.java:342)
>         at org.jboss.serial.persister.RegularObjectPersister.readData(RegularObjectPersister.java:239)
>         at org.jboss.serial.objectmetamodel.ObjectDescriptorFactory.readObjectDescriptionFromStreaming(ObjectDescriptorFactory.java:412)
>         at org.jboss.serial.objectmetamodel.ObjectDescriptorFactory.objectFromDescription(ObjectDescriptorFactory.java:82)
>         at org.jboss.serial.objectmetamodel.DataContainer$DataContainerDirectInput.readObject(DataContainer.java:643)
>         at org.jboss.serial.persister.RegularObjectPersister.readSlotWithFields(RegularObjectPersister.java:353)
>         at org.jboss.serial.persister.RegularObjectPersister.defaultRead(RegularObjectPersister.java:273)
>         at org.jboss.serial.persister.RegularObjectPersister.readData(RegularObjectPersister.java:241)
>         at org.jboss.serial.objectmetamodel.ObjectDescriptorFactory.readObjectDescriptionFromStreaming(ObjectDescriptorFactory.java:412)
>         at org.jboss.serial.objectmetamodel.ObjectDescriptorFactory.objectFromDescription(ObjectDescriptorFactory.java:82)
>         at org.jboss.serial.objectmetamodel.DataContainer$DataContainerDirectInput.readObject(DataContainer.java:643)
>         at org.jboss.serial.persister.RegularObjectPersister.readSlotWithFields(RegularObjectPersister.java:353)
>         at org.jboss.serial.persister.RegularObjectPersister.defaultRead(RegularObjectPersister.java:273)
>         at org.jboss.serial.persister.RegularObjectPersister.readData(RegularObjectPersister.java:241)
>         at org.jboss.serial.objectmetamodel.ObjectDescriptorFactory.readObjectDescriptionFromStreaming(ObjectDescriptorFactory.java:412)
>         at org.jboss.serial.objectmetamodel.ObjectDescriptorFactory.objectFromDescription(ObjectDescriptorFactory.java:82)
>         at org.jboss.serial.objectmetamodel.DataContainer$DataContainerDirectInput.readObject(DataContainer.java:643)
>         at org.jboss.serial.persister.RegularObjectPersister.readSlotWithFields(RegularObjectPersister.java:353)
>         at org.jboss.serial.persister.RegularObjectPersister.defaultRead(RegularObjectPersister.java:273)
>         at org.jboss.serial.persister.RegularObjectPersister.readData(RegularObjectPersister.java:241)
>         at org.jboss.serial.objectmetamodel.ObjectDescriptorFactory.readObjectDescriptionFromStreaming(ObjectDescriptorFactory.java:412)
>         at org.jboss.serial.objectmetamodel.ObjectDescriptorFactory.objectFromDescription(ObjectDescriptorFactory.java:82)
>         at org.jboss.serial.objectmetamodel.DataContainer$DataContainerDirectInput.readObject(DataContainer.java:643)
>         at org.jboss.serial.persister.RegularObjectPersister.readSlotWithFields(RegularObjectPersister.java:353)
>         at org.jboss.serial.persister.RegularObjectPersister.defaultRead(RegularObjectPersister.java:273)
>         at org.jboss.serial.persister.RegularObjectPersister.readData(RegularObjectPersister.java:241)
>         at org.jboss.serial.objectmetamodel.ObjectDescriptorFactory.readObjectDescriptionFromStreaming(ObjectDescriptorFactory.java:412)
> at org.jboss.serial.objectmetamodel.ObjectDescriptorFactory.objectFromDescription(ObjectDescriptorFactory.java:82)
>         at org.jboss.serial.objectmetamodel.DataContainer$DataContainerDirectInput.readObject(DataContainer.java:643)
>         at org.jboss.serial.io.JBossObjectInputStream.readObjectOverride(JBossObjectInputStream.java:163)
>         at java.io.ObjectInputStream.readObject(ObjectInputStream.java:365)
>         at com.griddynamics.jagger.util.SerializationUtils.deserialize(SerializationUtils.java:155)
>         ... 5 more
> Caused by: java.lang.InstantiationException: com.griddynamics.jagger.user.test.configurations.JTestDefinition$Builder
>         at java.lang.Class.newInstance(Class.java:427)
>         at org.jboss.serial.classmetamodel.ClassMetaData.newInstance(ClassMetaData.java:334)
>         ... 36 more
> Caused by: java.lang.NoSuchMethodException: com.griddynamics.jagger.user.test.configurations.JTestDefinition$Builder.<init>()
>         at java.lang.Class.getConstructor0(Class.java:3082)
>         at java.lang.Class.newInstance(Class.java:412)

**Comment by dlatnikov** [comment]
@tanoshko Tanya, please provide code sample that fails

**Comment by tanoshko** [comment]
@dlatnikov Dmitry, please see updated description.

**Comment by Amacuse** [comment]
The same exception is throwing when launching sample tests from: 

https://nexus.griddynamics.net/nexus/content/repositories/jagger-releases/

jagger.load.scenario.id.to.execute=ls_example
jagger.load.scenario.id.to.execute=myFirstJaggerLoadScenario

**Comment by Amacuse** [comment]
After our investigation we have discovered that the reason of the org.jboss.serial.exception.SerializationException is in a trait of serialization library:
  
https://mvnrepository.com/artifact/jboss/jboss-serialization/4.2.2.GA

that force classes, which will be serialized, to have a constructor without arguments. Classes that are serialized and sent to kernels:

- LoadBalancer
- EndpointsProvider
- QueryProvider
- Invoker
- Validator
- Listener 

All these classes and their members must have constructors with no arguments. This concerns custom and jagger implementations. But following jagger classes which implement load balancing behavior are not have no arguments constructors:

- CircularExclusiveAccessLoadBalancer
- NonCircularExclusiveAccessLoadBalancer
- RandomLoadBalancer(default)
- SimpleCircularLoadBalancer

and only two of them have it:

- OneByOneLoadBalancer
- RoundRobinLoadBalancer

Thus only them could be used in a distributed mode.

**Comment by tanoshko** [comment]
seems like this serializations library is quite old and isn't being updated. Please consider using some newer one.

**Comment by pzas03** [comment]
Issue was in the load balancer instantiation:
```
public class JTestDefinition {
...
    public static class Builder {
    ...
    	private Builder(Id id, Iterable endpointsProvider) {
    	...
	    this.loadBalancer = new RandomLoadBalancer() {{
	        setPairSupplierFactory(new RoundRobinPairSupplierFactory());
	        setRandomSeed(31);
	    }};
...
}
```
which was serialized as *com.griddynamics.jagger.user.test.configurations.JTestDefinition$Builder$1*

This is has been fixed in **2.0.1-m1**

**Comment by tanoshko** [comment]
Ok, thank you, closing this issue.

Let's consider improvement https://github.com/griddynamics/jagger/issues/849.
